distributed/tests/test_cancelled_state.py

```diff
@@ -561,3 +567,129 @@ def test_resume_executing_worker_state(ws_with_running_task):
+    assert instructions[1] == Execute(key="y", stimulus_id="s3")
+    assert ws.tasks["x"].state == "memory"
+    assert ws.data["x"] == 123
+
+
+@pytest.mark.xfail(reason="distributed#6565, distributed#6689")
+def test_workerstate_executing_failure_to_fetch(ws_with_running_task):
+    """Test state loops:
+
+    - executing -> cancelled -> resumed (fetch)
+    - executing -> long-running -> cancelled -> resumed (fetch)
+
+    The task execution later terminates with a failure.
+    This is an edge case interaction between work stealing and a task that does not
+    deterministically succeed or fail when run multiple times or on different workers.
+
+    Test that the task is fetched from the other worker. This is to avoid having to deal
+    with cancelling the dependent, which would require interaction with the scheduler
+    and increase the complexity of the use case.
+
+    See also: test_workerstate_executing_success_to_fetch
+    """
+    ws = ws_with_running_task
+    ws2 = "127.0.0.1:2"
+    instructions = ws.handle_stimulus(
+        FreeKeysEvent(keys=["x"], stimulus_id="s1"),
+        ComputeTaskEvent.dummy("y", who_has={"x": [ws2]}, stimulus_id="s2"),
+        ExecuteFailureEvent.dummy("x", stimulus_id="s3"),
+    )
+    assert instructions == [
+        GatherDep(worker=ws2, to_gather={"x"}, total_nbytes=1, stimulus_id="s3")
+    ]
+    assert ws.tasks["x"].state == "flight"
+
+
+def test_workerstate_flight_success_to_executing(ws):
+    """Test state loop
+
+    flight -> cancelled -> resumed (waiting)
+
+    gather_dep later terminates successfully.
+    Test that the task is not executed and is in memory.
+    """
+    ws2 = "127.0.0.1:2"
+    instructions = ws.handle_stimulus(
+        ComputeTaskEvent.dummy("y", who_has={"x": [ws2]}, stimulus_id="s1"),
+        FreeKeysEvent(keys=["y", "x"], stimulus_id="s2"),
+        ComputeTaskEvent.dummy("x", stimulus_id="s3"),
+        GatherDepSuccessEvent(
+            worker=ws2, total_nbytes=1, data={"x": 123}, stimulus_id="s4"
+        ),
+    )
+    assert len(instructions) == 2
+    assert instructions[0] == GatherDep(
+        worker=ws2, to_gather={"x"}, total_nbytes=1, stimulus_id="s1"
+    )
+    assert isinstance(instructions[1], TaskFinishedMsg)
+    assert instructions[1].stimulus_id == "s4"
+    assert ws.tasks["x"].state == "memory"
+    assert ws.data["x"] == 123
+
+
+def test_workerstate_flight_failure_to_executing(ws):
+    """Test state loop
+
+    flight -> cancelled -> resumed (waiting)
+
+    gather_dep later terminates with a failure.
+    Test that the task is executed.
+    """
+    ws2 = "127.0.0.1:2"
+    instructions = ws.handle_stimulus(
+        ComputeTaskEvent.dummy("y", who_has={"x": [ws2]}, stimulus_id="s1"),
+        FreeKeysEvent(keys=["y", "x"], stimulus_id="s2"),
+        ComputeTaskEvent.dummy("x", stimulus_id="s3"),
+        # Peer worker does not have the data
+        GatherDepSuccessEvent(worker=ws2, total_nbytes=1, data={}, stimulus_id="s4"),
```

Comment:
Shouldn't this be a `GatherDepFailureEvent`?

Reply:
A task can get out of flight in one of the following ways:

- GatherDepSuccessEvent, and the task is in data
- GatherDepSuccessEvent, but the task is not in data (the remote worker says it doesn't have it)
- GatherDepNetworkFailureEvent (the remote worker is dead)
- GatherDepBusyEvent (the remote worker is busy)
- GatherDepFailureEvent (edge case, generic error e.g. failure to deserialize)

All but the first are a failure to gather the key. That's what I was explaining in the commend on the previous line.

Reply:
From the docstring `gather_dep later terminates with a failure` I was under the impression that we were testing an actual `GatherDepFailureEvent`. Is there a way we can make this clearer? `gather_dep terminates successfully but does not have the data, resulting in a failure` might be an alternative to stress the point and the rather intricate difference between the successful event and the failing outcome.

Alternatively, I'd prefer just using anything but `GatherDepSuccessEvent` to avoid further confusion.